### ISSUE TITLE
Add option 'public' to cache providers

### ISSUE THREAD

**Issue by IndraGunawan** [doctrine/DoctrineCacheBundle#101]
in case cache provider only use on other services argument, so i dont need that provider set as public service.

for example i use doctrine cache for caching the validation metadata

``` yml
# service.yml
services:
    app.cache.validation:
        class: "Symfony\\Component\\Validator\\Mapping\\Cache\\DoctrineCache"
        arguments: ["@doctrine_cache.providers.validation_cache"]
        public: false

# config.yml
doctrine_cache:
    providers:
        validation_cache:
            namespace: "validation_cache:"
            redis:
                host: "..."
                port: "..."
                database: "..."

framework:
    validation:
        cache: app.cache.validation
```

the service `doctrine_cache.providers.validation_cache` just as other service arguments.

so i think it would be great if it can set as 'public' option

``` yml
# config.yml
doctrine_cache:
    providers:
        validation_cache:
            namespace: "validation_cache:"
            redis:
                host: "..."
                port: "..."
                database: "..."
            public: false
```

**Comment by mikeSimonson** [comment]
@IndraGunawan They are already made private by default.

https://github.com/doctrine/DoctrineCacheBundle/search?utf8=%E2%9C%93&q=setPublic%28false%29%3B

Regards

**Comment by IndraGunawan** [comment]
hi @mikeSimonson 

right, the connection service is private. but the cache provider services are public.

this is the result when i ran debug:container

```bash
[container] Information for service doctrine_cache.providers.default_cache
Service Id       doctrine_cache.providers.default_cache
Class            Doctrine\Common\Cache\RedisCache
Tags             -
Scope            container
Public           yes
Synthetic        no
Lazy             no
Synchronized     no
Abstract         no
```

**Comment by IndraGunawan** [comment]
how about if i create this feature?

**Comment by mikeSimonson** [comment]
It might be a better idea to ask @stof about it first. Just to make sure that it wouldn't break something else.

**Comment by mikeSimonson** [comment]
@IndraGunawan It looks like [there is a test that verify that the cache provider is public](https://github.com/doctrine/DoctrineCacheBundle/blob/master/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php#L332). But I can't find any place where that test assertion is called ...

**Comment by Tobion** [comment]
As the default for "public" changes in Symfony 4, all cache providers are private when used with SF 4. They are not explicitly set to public in https://github.com/doctrine/DoctrineCacheBundle/blob/master/DependencyInjection/CacheProviderLoader.php (neither are the aliases).

We're getting deprecations in SF 3.4 due to the providers going to be private. I'm fine if they are private. So should we just not do anything and let them transition to private or set them to public or private explicitly? @stof what's your opinion?

**Comment by nacc** [comment]
Any word on this? It seems like this fails the unit tests on Ubuntu 18.04:

```
There were 10 failures:

1) Doctrine\Bundle\DoctrineCacheBundle\Tests\DependencyInjection\XmlDoctrineCacheExtensionTest::testBasicCache
Failed asserting that false is true.

/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:317
/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:87

2) Doctrine\Bundle\DoctrineCacheBundle\Tests\DependencyInjection\XmlDoctrineCacheExtensionTest::testBasicConfigurableCache
Failed asserting that false is true.

/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:317
/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:128

3) Doctrine\Bundle\DoctrineCacheBundle\Tests\DependencyInjection\XmlDoctrineCacheExtensionTest::testBasicConfigurableDefaultCache
Failed asserting that false is true.

/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:317
/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:187

4) Doctrine\Bundle\DoctrineCacheBundle\Tests\DependencyInjection\XmlDoctrineCacheExtensionTest::testServiceParameters
Failed asserting that false is true.

/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:317
/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:259

5) Doctrine\Bundle\DoctrineCacheBundle\Tests\DependencyInjection\XmlDoctrineCacheExtensionTest::testCustomCacheProviders
Failed asserting that false is true.

/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:317
/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:281

6) Doctrine\Bundle\DoctrineCacheBundle\Tests\DependencyInjection\YmlDoctrineCacheExtensionTest::testBasicCache
Failed asserting that false is true.

/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:317
/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:87

7) Doctrine\Bundle\DoctrineCacheBundle\Tests\DependencyInjection\YmlDoctrineCacheExtensionTest::testBasicConfigurableCache
Failed asserting that false is true.

/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:317
/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:128

8) Doctrine\Bundle\DoctrineCacheBundle\Tests\DependencyInjection\YmlDoctrineCacheExtensionTest::testBasicConfigurableDefaultCache
Failed asserting that false is true.

/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:317
/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:187

9) Doctrine\Bundle\DoctrineCacheBundle\Tests\DependencyInjection\YmlDoctrineCacheExtensionTest::testServiceParameters
Failed asserting that false is true.

/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:317
/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:259

10) Doctrine\Bundle\DoctrineCacheBundle\Tests\DependencyInjection\YmlDoctrineCacheExtensionTest::testCustomCacheProviders
Failed asserting that false is true.

/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:317
/build/php-doctrine-cache-bundle-lNz4HJ/php-doctrine-cache-bundle-1.3.2/Tests/DependencyInjection/AbstractDoctrineCacheExtensionTest.php:281
```

Using 1.3.2 and Symfony 3.4.3.

**Comment by Tschebel** [comment]
I've noticed an additional problem with private cache providers.

The commands doctrine:cache:stats / clear / ... get the provider name as an argument and than try to fetch it from the container which is not possible with private services.

**Comment by Tobion** [comment]
I can confirm @Tschebel

**Comment by Tobion** [comment]
@fabpot this can be closed with #135
